src/workerd/api/sockets.c++

```diff
@@ -164,23 +164,34 @@ jsg::Ref<Socket> Socket::startTls(jsg::Lock& js, jsg::Optional<TlsOptions> tlsOp
 
   // The current socket's writable buffers need to be flushed. The socket's WritableStream is backed
   // by an AsyncIoStream which doesn't implement any buffering, so we don't need to worry about
-  // flushing. But this is something to keep in mind in case this assumption no longer holds in
-  // the future.
+  // flushing. But the JS WritableStream holds a queue so some data may still be buffered. This
+  // means we need to flush the WritableStream.
   //
   // Detach the AsyncIoStream from the Writable/Readable streams and make them unusable.
-  writable->removeSink(js);
-  readable = readable->detach(js, true);
-  closeFulfiller.resolver.resolve();
-
-  auto acceptedHostname = domain.asPtr();
-  KJ_IF_MAYBE(s, tlsOptions) {
-    KJ_IF_MAYBE(expectedHost, s->expectedServerHostname) {
-      acceptedHostname = *expectedHost;
+  auto& context = IoContext::current();
+  auto secureStreamPromise = context.awaitJs(writable->flush(js).then(js,
```

Comment:
What happens here if the writable errors while trying to flush?

Reply:
I would hope that it gets propagated up as part of the PromisedAsyncIoStream below. Is that a fair assumption?

Reply:
it should. let's be sure to test that case in the starttls tests just to make sure :-)

Reply:
Alright, will add to my TODOs.